iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.inl

```diff
@@ -25,23 +25,60 @@ namespace iox
 namespace rp
 {
 template <typename T>
-inline RelativePointer<T>::RelativePointer(ptr_t ptr, id_t id) noexcept
-    : BaseRelativePointer(ptr, id)
+// NOLINTJUSTIFICATION NewType size is comparable to an integer, hence pass by value is preferred
+// NOLINTNEXTLINE(performance-unnecessary-value-param)
+inline RelativePointer<T>::RelativePointer(ptr_t ptr, segment_id_t id) noexcept
+    : m_id(id)
+    , m_offset(computeOffset(ptr))
 {
 }
 
 template <typename T>
-inline RelativePointer<T>::RelativePointer(offset_t offset, id_t id) noexcept
-    : BaseRelativePointer(offset, id)
+// NOLINTJUSTIFICATION NewType size is comparable to an integer, hence pass by value is preferred
+// NOLINTNEXTLINE(performance-unnecessary-value-param)
+inline RelativePointer<T>::RelativePointer(offset_t offset, segment_id_t id) noexcept
+    : m_id(id)
+    , m_offset(offset)
 {
 }
 
 template <typename T>
 inline RelativePointer<T>::RelativePointer(ptr_t ptr) noexcept
-    : BaseRelativePointer(ptr)
+    : m_id(searchId(ptr))
+    , m_offset(computeOffset(ptr))
 {
 }
 
+template <typename T>
+inline RelativePointer<T>& RelativePointer<T>::operator=(const RelativePointer& other) noexcept
+{
+    if (this != &other)
+    {
+        m_id = other.m_id;
+        m_offset = other.m_offset;
+    }
+    return *this;
+}
+
+template <typename T>
+RelativePointer<T>::RelativePointer(RelativePointer&& other) noexcept
+{
+    *this = std::move(other);
```

Comment:
@MatthiasKillat Please enlighten me :grin:

Reply:
1. This way we have the self assignment assignment check here (but self-assignment cannot happen in the ctor!)
2. We default initialize and then immediately assign (the compiler cannot prevent this in all cases as a member CTOR is programmer intention). This is mainly an issue if there are non primitive types, more so if they are large.

Short story is that the correct way is to use the initializer list to initilaize the members (if they depend on ctor args). To reduce code duplication one can either use a helper function or implement assignment in terms of copy via copy-and-swap idiom.

The code as is is just suboptimal and not the idiomatic C++ way.

Reply:
We do not have much of an issue here, just the annoying self assignment check. I would actually prefer to initialize the members directly in the initializer list and invalidate other in the body (if we care for this, it is not needed as with unique_ptr).

Reply:
As this is general topic and lots of classes are affected, could you please copy the above to a new issue and we continue to discuss it there?

Reply:
It is already inconsistent among classes and this does not matter too much. It is very simple to change here, like 4 lines of code  at most. I would not create an issue for this now, as I do not think we should refactor all of them right away (only for new classes or if we refactor for another reason).

Reply:
@MatthiasKillat Ok, I will change it on the follow-up PR #1677. As I think this is an important finding to document for our future selves to avoid discussing this again, could you please create this issue?

Reply:
@MatthiasKillat @FerdinandSpitzschnueffler I created: #1686